<a href="https://colab.research.google.com/github/pavannaik321/Brest_cancer_detection/blob/main/BrestCancer.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**We are taking data from uci datasets**

Ignore all warnings which are printing again and again

In [179]:
import warnings
warnings.filterwarnings('ignore')

In [180]:
# from google.colab import drive
# drive.mount('/content/drive')

In [181]:
import pandas as pd

import numpy as np


datasets are from UCI repository

In [182]:
path = 'drive/MyDrive/AI projects/Breast cancer';

In [183]:
df = pd.read_csv(path+"/wdbc.data")
df

842302  M  17.99  10.38   122.8    1001   0.1184   0.2776   0.3001  \
0      842517  M  20.57  17.77  132.90  1326.0  0.08474  0.07864  0.08690   
1    84300903  M  19.69  21.25  130.00  1203.0  0.10960  0.15990  0.19740   
2    84348301  M  11.42  20.38   77.58   386.1  0.14250  0.28390  0.24140   
3    84358402  M  20.29  14.34  135.10  1297.0  0.10030  0.13280  0.19800   
4      843786  M  12.45  15.70   82.57   477.1  0.12780  0.17000  0.15780   
..        ... ..    ...    ...     ...     ...      ...      ...      ...   
563    926424  M  21.56  22.39  142.00  1479.0  0.11100  0.11590  0.24390   
564    926682  M  20.13  28.25  131.20  1261.0  0.09780  0.10340  0.14400   
565    926954  M  16.60  28.08  108.30   858.1  0.08455  0.10230  0.09251   
566    927241  M  20.60  29.33  140.10  1265.0  0.11780  0.27700  0.35140   
567     92751  B   7.76  24.54   47.92   181.0  0.05263  0.04362  0.00000   

      0.1471  ...   25.38  17.33   184.6    2019   0.1622   0.6656  0.7119  \
0    0.07017  ...  24.990  23.41  158.80  1956.0  0.12380  0.18660  0.2416   
1    0.12790  ...  23.570  25.53  152.50  1709.0  0.14440  0.42450  0.4504   
2    0.10520  ...  14.910  26.50   98.87   567.7  0.20980  0.86630  0.6869   
3    0.10430  ...  22.540  16.67  152.20  1575.0  0.13740  0.20500  0.4000   
4    0.08089  ...  15.470  23.75  103.40   741.6  0.17910  0.52490  0.5355   
..       ...  ...     ...    ...     ...     ...      ...      ...     ...   
563  0.13890  ...  25.450  26.40  166.10  2027.0  0.14100  0.21130  0.4107   
564  0.09791  ...  23.690  38.25  155.00  1731.0  0.11660  0.19220  0.3215   
565  0.05302  ...  18.980  34.12  126.70  1124.0  0.11390  0.30940  0.3403   
566  0.15200  ...  25.740  39.42  184.60  1821.0  0.16500  0.86810  0.9387   
567  0.00000  ...   9.456  30.37   59.16   268.6  0.08996  0.06444  0.0000   

     0.2654  0.4601   0.1189  
0    0.1860  0.2750  0.08902  
1    0.2430  0.3613  0.08758  
2    0.2575  0.6638  0.17300  
3    0.1625  0.2364  0.07678  
4    0.1741  0.3985  0.12440  
..      ...     ...      ...  
563  0.2216  0.2060  0.07115  
564  0.1628  0.2572  0.06637  
565  0.1418  0.2218  0.07820  
566  0.2650  0.4087  0.12400  
567  0.0000  0.2871  0.07039  

[568 rows x 32 columns]

#data cleaning

In [184]:
df.drop(['842302'], axis=1 , inplace = True)

In [185]:
df

M  17.99  10.38   122.8    1001   0.1184   0.2776   0.3001   0.1471  \
0    M  20.57  17.77  132.90  1326.0  0.08474  0.07864  0.08690  0.07017   
1    M  19.69  21.25  130.00  1203.0  0.10960  0.15990  0.19740  0.12790   
2    M  11.42  20.38   77.58   386.1  0.14250  0.28390  0.24140  0.10520   
3    M  20.29  14.34  135.10  1297.0  0.10030  0.13280  0.19800  0.10430   
4    M  12.45  15.70   82.57   477.1  0.12780  0.17000  0.15780  0.08089   
..  ..    ...    ...     ...     ...      ...      ...      ...      ...   
563  M  21.56  22.39  142.00  1479.0  0.11100  0.11590  0.24390  0.13890   
564  M  20.13  28.25  131.20  1261.0  0.09780  0.10340  0.14400  0.09791   
565  M  16.60  28.08  108.30   858.1  0.08455  0.10230  0.09251  0.05302   
566  M  20.60  29.33  140.10  1265.0  0.11780  0.27700  0.35140  0.15200   
567  B   7.76  24.54   47.92   181.0  0.05263  0.04362  0.00000  0.00000   

     0.2419  ...   25.38  17.33   184.6    2019   0.1622   0.6656  0.7119  \
0    0.1812  ...  24.990  23.41  158.80  1956.0  0.12380  0.18660  0.2416   
1    0.2069  ...  23.570  25.53  152.50  1709.0  0.14440  0.42450  0.4504   
2    0.2597  ...  14.910  26.50   98.87   567.7  0.20980  0.86630  0.6869   
3    0.1809  ...  22.540  16.67  152.20  1575.0  0.13740  0.20500  0.4000   
4    0.2087  ...  15.470  23.75  103.40   741.6  0.17910  0.52490  0.5355   
..      ...  ...     ...    ...     ...     ...      ...      ...     ...   
563  0.1726  ...  25.450  26.40  166.10  2027.0  0.14100  0.21130  0.4107   
564  0.1752  ...  23.690  38.25  155.00  1731.0  0.11660  0.19220  0.3215   
565  0.1590  ...  18.980  34.12  126.70  1124.0  0.11390  0.30940  0.3403   
566  0.2397  ...  25.740  39.42  184.60  1821.0  0.16500  0.86810  0.9387   
567  0.1587  ...   9.456  30.37   59.16   268.6  0.08996  0.06444  0.0000   

     0.2654  0.4601   0.1189  
0    0.1860  0.2750  0.08902  
1    0.2430  0.3613  0.08758  
2    0.2575  0.6638  0.17300  
3    0.1625  0.2364  0.07678  
4    0.1741  0.3985  0.12440  
..      ...     ...      ...  
563  0.2216  0.2060  0.07115  
564  0.1628  0.2572  0.06637  
565  0.1418  0.2218  0.07820  
566  0.2650  0.4087  0.12400  
567  0.0000  0.2871  0.07039  

[568 rows x 31 columns]

check how many datasets we have for malignant and benign

In [186]:
df['M'].value_counts()

B    357
M    211
Name: M, dtype: int64

211 are affected by brest cancer and 357 are not affected by brest cancer

#feature selection

lean various feature selection technique

In [187]:
from sklearn.model_selection import train_test_split

convert to categorical value

In [188]:
diag_map = {'M' : 1 , 'B' : 0}
df['M'] = df['M'].map(diag_map)

In [189]:
df

M  17.99  10.38   122.8    1001   0.1184   0.2776   0.3001   0.1471  \
0    1  20.57  17.77  132.90  1326.0  0.08474  0.07864  0.08690  0.07017   
1    1  19.69  21.25  130.00  1203.0  0.10960  0.15990  0.19740  0.12790   
2    1  11.42  20.38   77.58   386.1  0.14250  0.28390  0.24140  0.10520   
3    1  20.29  14.34  135.10  1297.0  0.10030  0.13280  0.19800  0.10430   
4    1  12.45  15.70   82.57   477.1  0.12780  0.17000  0.15780  0.08089   
..  ..    ...    ...     ...     ...      ...      ...      ...      ...   
563  1  21.56  22.39  142.00  1479.0  0.11100  0.11590  0.24390  0.13890   
564  1  20.13  28.25  131.20  1261.0  0.09780  0.10340  0.14400  0.09791   
565  1  16.60  28.08  108.30   858.1  0.08455  0.10230  0.09251  0.05302   
566  1  20.60  29.33  140.10  1265.0  0.11780  0.27700  0.35140  0.15200   
567  0   7.76  24.54   47.92   181.0  0.05263  0.04362  0.00000  0.00000   

     0.2419  ...   25.38  17.33   184.6    2019   0.1622   0.6656  0.7119  \
0    0.1812  ...  24.990  23.41  158.80  1956.0  0.12380  0.18660  0.2416   
1    0.2069  ...  23.570  25.53  152.50  1709.0  0.14440  0.42450  0.4504   
2    0.2597  ...  14.910  26.50   98.87   567.7  0.20980  0.86630  0.6869   
3    0.1809  ...  22.540  16.67  152.20  1575.0  0.13740  0.20500  0.4000   
4    0.2087  ...  15.470  23.75  103.40   741.6  0.17910  0.52490  0.5355   
..      ...  ...     ...    ...     ...     ...      ...      ...     ...   
563  0.1726  ...  25.450  26.40  166.10  2027.0  0.14100  0.21130  0.4107   
564  0.1752  ...  23.690  38.25  155.00  1731.0  0.11660  0.19220  0.3215   
565  0.1590  ...  18.980  34.12  126.70  1124.0  0.11390  0.30940  0.3403   
566  0.2397  ...  25.740  39.42  184.60  1821.0  0.16500  0.86810  0.9387   
567  0.1587  ...   9.456  30.37   59.16   268.6  0.08996  0.06444  0.0000   

     0.2654  0.4601   0.1189  
0    0.1860  0.2750  0.08902  
1    0.2430  0.3613  0.08758  
2    0.2575  0.6638  0.17300  
3    0.1625  0.2364  0.07678  
4    0.1741  0.3985  0.12440  
..      ...     ...      ...  
563  0.2216  0.2060  0.07115  
564  0.1628  0.2572  0.06637  
565  0.1418  0.2218  0.07820  
566  0.2650  0.4087  0.12400  
567  0.0000  0.2871  0.07039  

[568 rows x 31 columns]

In [190]:
#X = df.iloc[: , 1:-1]
X = df.loc[: , df.columns!='M']

In [191]:
X

17.99  10.38   122.8    1001   0.1184   0.2776   0.3001   0.1471  0.2419  \
0    20.57  17.77  132.90  1326.0  0.08474  0.07864  0.08690  0.07017  0.1812   
1    19.69  21.25  130.00  1203.0  0.10960  0.15990  0.19740  0.12790  0.2069   
2    11.42  20.38   77.58   386.1  0.14250  0.28390  0.24140  0.10520  0.2597   
3    20.29  14.34  135.10  1297.0  0.10030  0.13280  0.19800  0.10430  0.1809   
4    12.45  15.70   82.57   477.1  0.12780  0.17000  0.15780  0.08089  0.2087   
..     ...    ...     ...     ...      ...      ...      ...      ...     ...   
563  21.56  22.39  142.00  1479.0  0.11100  0.11590  0.24390  0.13890  0.1726   
564  20.13  28.25  131.20  1261.0  0.09780  0.10340  0.14400  0.09791  0.1752   
565  16.60  28.08  108.30   858.1  0.08455  0.10230  0.09251  0.05302  0.1590   
566  20.60  29.33  140.10  1265.0  0.11780  0.27700  0.35140  0.15200  0.2397   
567   7.76  24.54   47.92   181.0  0.05263  0.04362  0.00000  0.00000  0.1587   

     0.07871  ...   25.38  17.33   184.6    2019   0.1622   0.6656  0.7119  \
0    0.05667  ...  24.990  23.41  158.80  1956.0  0.12380  0.18660  0.2416   
1    0.05999  ...  23.570  25.53  152.50  1709.0  0.14440  0.42450  0.4504   
2    0.09744  ...  14.910  26.50   98.87   567.7  0.20980  0.86630  0.6869   
3    0.05883  ...  22.540  16.67  152.20  1575.0  0.13740  0.20500  0.4000   
4    0.07613  ...  15.470  23.75  103.40   741.6  0.17910  0.52490  0.5355   
..       ...  ...     ...    ...     ...     ...      ...      ...     ...   
563  0.05623  ...  25.450  26.40  166.10  2027.0  0.14100  0.21130  0.4107   
564  0.05533  ...  23.690  38.25  155.00  1731.0  0.11660  0.19220  0.3215   
565  0.05648  ...  18.980  34.12  126.70  1124.0  0.11390  0.30940  0.3403   
566  0.07016  ...  25.740  39.42  184.60  1821.0  0.16500  0.86810  0.9387   
567  0.05884  ...   9.456  30.37   59.16   268.6  0.08996  0.06444  0.0000   

     0.2654  0.4601   0.1189  
0    0.1860  0.2750  0.08902  
1    0.2430  0.3613  0.08758  
2    0.2575  0.6638  0.17300  
3    0.1625  0.2364  0.07678  
4    0.1741  0.3985  0.12440  
..      ...     ...      ...  
563  0.2216  0.2060  0.07115  
564  0.1628  0.2572  0.06637  
565  0.1418  0.2218  0.07820  
566  0.2650  0.4087  0.12400  
567  0.0000  0.2871  0.07039  

[568 rows x 30 columns]

In [192]:
Y = df[['M']]

In [193]:
Y

M
0    1
1    1
2    1
3    1
4    1
..  ..
563  1
564  1
565  1
566  1
567  0

[568 rows x 1 columns]

#Spilit into training and testing datasets

In [194]:
X_train , X_test , Y_train , Y_test = train_test_split(X,Y,test_size = 0.2,random_state = 42)

In [195]:
X_train

17.99  10.38   122.8   1001   0.1184   0.2776   0.3001   0.1471  0.2419  \
248  11.520  14.93   73.87  406.3  0.10130  0.07808  0.04328  0.02929  0.1883   
88   14.640  15.24   95.77  651.9  0.11320  0.13390  0.09966  0.07064  0.2116   
334  17.060  21.00  111.80  918.6  0.11190  0.10560  0.15080  0.09934  0.1727   
362  16.500  18.29  106.60  838.1  0.09686  0.08468  0.05862  0.04835  0.1495   
33   16.130  17.88  107.00  807.2  0.10400  0.15590  0.13540  0.07752  0.1998   
..      ...    ...     ...    ...      ...      ...      ...      ...     ...   
71   17.200  24.52  114.20  929.4  0.10710  0.18300  0.16920  0.07944  0.1927   
106  12.360  18.54   79.01  466.7  0.08477  0.06815  0.02643  0.01921  0.1602   
270  11.290  13.04   72.23  388.0  0.09834  0.07608  0.03265  0.02755  0.1769   
435  12.870  19.54   82.67  509.2  0.09136  0.07883  0.01797  0.02090  0.1861   
102   9.876  19.40   63.95  298.3  0.10050  0.09697  0.06154  0.03029  0.1945   

     0.07871  ...  25.38  17.33   184.6    2019  0.1622  0.6656   0.7119  \
248  0.06168  ...  12.65  21.19   80.88   491.8  0.1389  0.1582  0.18040   
88   0.06346  ...  16.34  18.24  109.40   803.6  0.1277  0.3089  0.26040   
334  0.06071  ...  20.99  33.15  143.20  1362.0  0.1449  0.2053  0.39200   
362  0.05593  ...  18.13  25.45  117.20  1009.0  0.1338  0.1679  0.16630   
33   0.06515  ...  20.21  27.26  132.70  1261.0  0.1446  0.5804  0.52740   
..       ...  ...    ...    ...     ...     ...     ...     ...      ...   
71   0.06487  ...  23.32  33.82  151.60  1681.0  0.1585  0.7394  0.65660   
106  0.06066  ...  13.29  27.49   85.56   544.1  0.1184  0.1963  0.19370   
270  0.06270  ...  12.32  16.18   78.27   457.5  0.1358  0.1507  0.12750   
435  0.06347  ...  14.45  24.38   95.14   626.9  0.1214  0.1652  0.07127   
102  0.06322  ...  10.76  26.83   72.22   361.2  0.1559  0.2302  0.26440   

      0.2654  0.4601   0.1189  
248  0.09608  0.2664  0.07809  
88   0.13970  0.3151  0.08473  
334  0.18270  0.2623  0.07599  
362  0.09123  0.2394  0.06469  
33   0.18640  0.4270  0.12330  
..       ...     ...      ...  
71   0.18990  0.3313  0.13390  
106  0.08442  0.2983  0.07185  
270  0.08750  0.2733  0.08022  
435  0.06384  0.3313  0.07735  
102  0.09749  0.2622  0.08490  

[454 rows x 30 columns]

In [196]:
X_test

17.99  10.38   122.8    1001   0.1184   0.2776   0.3001   0.1471  0.2419  \
218  19.53  32.47  128.00  1223.0  0.08420  0.11300  0.11450  0.06637  0.1428   
79   11.45  20.97   73.81   401.5  0.11020  0.09362  0.04591  0.02233  0.1842   
104  13.11  15.56   87.21   530.2  0.13980  0.17650  0.20710  0.09601  0.1925   
208  15.27  12.91   98.17   725.5  0.08182  0.06230  0.05892  0.03157  0.1359   
543  13.87  20.70   89.77   584.8  0.09578  0.10180  0.03688  0.02369  0.1620   
..     ...    ...     ...     ...      ...      ...      ...      ...     ...   
148  13.74  17.91   88.12   585.0  0.07944  0.06376  0.02881  0.01329  0.1473   
247  10.65  25.22   68.01   347.0  0.09657  0.07234  0.02379  0.01615  0.1897   
15   14.68  20.13   94.74   684.5  0.09867  0.07200  0.07395  0.05259  0.1586   
557  14.59  22.68   96.39   657.1  0.08473  0.13300  0.10290  0.03736  0.1454   
245  13.20  17.43   84.13   541.6  0.07215  0.04524  0.04336  0.01105  0.1487   

     0.07871  ...  25.38  17.33   184.6    2019   0.1622  0.6656  0.7119  \
218  0.05313  ...  27.90  45.41  180.20  2477.0  0.14080  0.4097  0.3995   
79   0.07005  ...  13.11  32.16   84.53   525.1  0.15570  0.1676  0.1755   
104  0.07692  ...  16.31  22.40  106.40   827.2  0.18620  0.4099  0.6376   
208  0.05526  ...  17.38  15.92  113.70   932.7  0.12220  0.2186  0.2962   
543  0.06688  ...  15.05  24.75   99.17   688.6  0.12640  0.2037  0.1377   
..       ...  ...    ...    ...     ...     ...      ...     ...     ...   
148  0.05580  ...  15.34  22.46   97.19   725.9  0.09711  0.1824  0.1564   
247  0.06329  ...  12.25  35.19   77.98   455.7  0.14990  0.1398  0.1125   
15   0.05922  ...  19.07  30.88  123.40  1138.0  0.14640  0.1871  0.2914   
557  0.06147  ...  15.48  27.27  105.90   733.5  0.10260  0.3171  0.3662   
245  0.05635  ...  13.94  27.82   88.28   602.0  0.11010  0.1508  0.2298   

      0.2654  0.4601   0.1189  
218  0.16250  0.2713  0.07568  
79   0.06127  0.2762  0.08851  
104  0.19860  0.3147  0.14050  
208  0.10350  0.2320  0.07474  
543  0.06845  0.2249  0.08492  
..       ...     ...      ...  
148  0.06019  0.2350  0.07014  
247  0.06136  0.3409  0.08147  
15   0.16090  0.3029  0.08216  
557  0.11050  0.2258  0.08004  
245  0.04970  0.2767  0.07198  

[114 rows x 30 columns]

In [197]:
Y_train

M
248  0
88   0
334  1
362  0
33   1
..  ..
71   1
106  0
270  0
435  0
102  0

[454 rows x 1 columns]

In [198]:
Y_test

M
218  1
79   0
104  1
208  0
543  0
..  ..
148  0
247  0
15   1
557  0
245  0

[114 rows x 1 columns]

#Model -KNN

In [199]:
from sklearn.neighbors import KNeighborsClassifier

In [200]:
knn = KNeighborsClassifier()

In [201]:
knn.fit(X_train,Y_train)

KNeighborsClassifier()

In [202]:
knn_Y_pred = knn.predict(X_test)

In [203]:
knn_Y_pred

array([1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1,
       0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0,
       0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1,
       0, 0, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0,
       0, 1, 0, 0])

We can find the accuracy also by using library

In [204]:
from sklearn.metrics import accuracy_score

In [205]:
accuracy_score(knn_Y_pred , Y_test)*100

92.10526315789474

#lets test some other models also
#Model - Logistic regression

In [206]:
from sklearn.linear_model import LogisticRegression

Creating instance of the model

In [207]:
lr = LogisticRegression(random_state = 0)

In [208]:
lr.fit(X_train , Y_train)

LogisticRegression(random_state=0)

In [209]:
LogisticRegression(random_state=0)
lr_y_pred = lr.predict(X_test)
accuracy_score(lr_y_pred , Y_test)*100

91.22807017543859

#Model - Naive Bayes

In [210]:
from sklearn.naive_bayes import GaussianNB


create instance

In [211]:
gnb = GaussianNB()

In [212]:
gnb.fit(X_train , Y_train);


In [213]:
gnb_y_pred = gnb.predict(X_test)

In [214]:
accuracy_score(gnb_y_pred , Y_test)*100

92.98245614035088

#Out of 3 models Naive Bayes works better than knn and logistic regression

#Model - K Fold Cross Validation

every time when we are shuffling the X_train and Y_train we are getting seprate accuracy so to avoid this we are using K Cross validation it will give us the mean of the accuracy's

In [215]:
from sklearn.model_selection import cross_val_score


below we are creating two variables first one will store all the accuracy and the second will store the cross validation accuracy

In [216]:
accuracy_all = []
cross_val_accuracy_all = []

the below function will take 3 parameters (model_name , X , Y , no of times we want to train our model )

#KNN

In [217]:
scores = cross_val_score(knn,X,Y,cv=10)
accuracy_all.append(accuracy_score(knn_Y_pred,Y_test))
cross_val_accuracy_all.append(np.mean(scores))

print("Accuracy : {0:.2%}".format(accuracy_score(knn_Y_pred, Y_test)))
print("10 fold Cross Validation Score : {0:.2%} (+/-  {1:.2%})".format(np.mean(scores),np.std(scores)*2))

Accuracy : 92.11%
Cross Validation Score : 92.96% (+/-  5.88%)


##Logistic Regression

In [219]:
accuracy_all = []
cross_val_accuracy_all = []

scores = cross_val_score(lr,X,Y,cv=10)
accuracy_all.append(accuracy_score(lr_y_pred,Y_test))
cross_val_accuracy_all.append(np.mean(scores))

print("Accuracy : {0:.2%}".format(accuracy_score(lr_y_pred, Y_test)))
print("10 fold Cross Validation Score : {0:.2%} (+/-  {1:.2%})".format(np.mean(scores),np.std(scores)*2))

Accuracy : 91.23%
Cross Validation Score : 94.19% (+/-  2.74%)


##Naive Byes


In [220]:
accuracy_all = []
cross_val_accuracy_all = []

scores = cross_val_score(gnb,X,Y,cv=10)
accuracy_all.append(accuracy_score(gnb_y_pred,Y_test))
cross_val_accuracy_all.append(np.mean(scores))

print("Accuracy : {0:.2%}".format(accuracy_score(gnb_y_pred, Y_test)))
print("10 fold Cross Validation Score : {0:.2%} (+/-  {1:.2%})".format(np.mean(scores),np.std(scores)*2))

Accuracy : 92.98%
Cross Validation Score : 94.02% (+/-  5.47%)
